### PR TITLE
[ASM][IAST] Configure maximum IAST Ranges

### PR DESCRIPTION
**Pull request by e-n-0**
## Summary of changes

Add the new environment variable `DD_IAST_MAX_RANGE_COUNT` defined [here](https://docs.google.com/document/d/13nP-3Uyz-HoDIhzUjqpfBABM8AjIwTuzQvc71R_oMEk/edit#heading=h.4mqpiqkued9p).

## Reason for change

- Required for IAST GA.
- Troubleshoot a client issue 

## Implementation details

In all source of creation of new ranges arrays, the limit is checked to prevent the creation of a larger Ranges array.
A RangeList object has been created to facilitate the addition of Ranges.

## Test coverage

- Unit tests: 
    - Test against `String` implementation (`concat`, `insert`)
    - Test against `StringBuilder` implementation (`append`)
- Integration tests:
    -  New endpoint `MaxRanges` in the `AspNetCore5` sample, where an arbitrary has been choose (`ReflectionInjection` here)  to see that the evidence value parts has been correctly written corresponding to the configuration.

### CODE REVIEW

**Comment by andrewlock** [comment]
## Execution-Time Benchmarks Report :stopwatch:

Execution-time results for samples comparing the following branches/commits:
- [This PR (5292)](https://github.com/DataDog/dd-trace-dotnet/tree/33636587056b392cecd65a4ae63439882bacb6f8)
- [master](https://github.com/DataDog/dd-trace-dotnet/tree/a8cb7e356d1aab3f3f4d0f170c00df0d5aaf7b9e)

Execution-time benchmarks measure the whole time it takes to execute a program. And are intended to measure the one-off costs. Cases where the execution time results for the PR are worse than latest master results are shown in **red**. The following thresholds were used for comparing the execution times:
* Welch test with statistical test for significance of **5%**
* Only results indicating a difference greater than **5%** and **5 ms** are considered.


Note that these results are based on a _single_ point-in-time result for each branch. For full results, see the [dashboard](https://ddstaging.datadoghq.com/dashboard/4qn-6fi-54p/apm-net-execution-time-benchmarks).

Graphs show the p99 interval based on the mean and StdDev of the test run, as well as the mean value of the run (shown as a diamond below the graph).
```mermaid
gantt
    title Execution time (ms) FakeDbCommand (.NET Framework 4.6.2) 
    dateFormat  X
    axisFormat %s
    todayMarker off
    section Baseline
    This PR (5292) - mean (76ms)  : 67, 85
     .   : milestone, 76,
    master - mean (74ms)  : 65, 83
     .   : milestone, 74,

    section CallTarget+Inlining+NGEN
    This PR (5292) - mean (988ms)  : 958, 1019
     .   : milestone, 988,
    master - mean (985ms)  : 961, 1008
     .   : milestone, 985,

```
```mermaid
gantt
    title Execution time (ms) FakeDbCommand (.NET Core 3.1) 
    dateFormat  X
    axisFormat %s
    todayMarker off
    section Baseline
    This PR (5292) - mean (111ms)  : 107, 115
     .   : milestone, 111,
    master - mean (118ms)  : 96, 140
     .   : milestone, 118,

    section CallTarget+Inlining+NGEN
    This PR (5292) - mean (716ms)  : 692, 739
     .   : milestone, 716,
    master - mean (713ms)  : 691, 735
     .   : milestone, 713,

```
```mermaid
gantt
    title Execution time (ms) FakeDbCommand (.NET 6) 
    dateFormat  X
    axisFormat %s
    todayMarker off
    section Baseline
    This PR (5292) - mean (95ms)  : 92, 98
     .   : milestone, 95,
    master - mean (95ms)  : 91, 98
     .   : milestone, 95,

    section CallTarget+Inlining+NGEN
    This PR (5292) - mean (670ms)  : 641, 700
     .   : milestone, 670,
    master - mean (671ms)  : 647, 694
     .   : milestone, 671,

```
```mermaid
gantt
    title Execution time (ms) HttpMessageHandler (.NET Framework 4.6.2) 
    dateFormat  X
    axisFormat %s
    todayMarker off
    section Baseline
    This PR (5292) - mean (188ms)  : 186, 191
     .   : milestone, 188,
    master - mean (188ms)  : 182, 193
     .   : milestone, 188,

    section CallTarget+Inlining+NGEN
    This PR (5292) - mean (1,073ms)  : 1046, 1099
     .   : milestone, 1073,
    master - mean (1,063ms)  : 1038, 1089
     .   : milestone, 1063,

```
```mermaid
gantt
    title Execution time (ms) HttpMessageHandler (.NET Core 3.1) 
    dateFormat  X
    axisFormat %s
    todayMarker off
    section Baseline
    This PR (5292) - mean (272ms)  : 267, 277
     .   : milestone, 272,
    master - mean (269ms)  : 262, 276
     .   : milestone, 269,

    section CallTarget+Inlining+NGEN
    This PR (5292) - mean (875ms)  : 851, 900
     .   : milestone, 875,
    master - mean (861ms)  : 839, 884
     .   : milestone, 861,

```
```mermaid
gantt
    title Execution time (ms) HttpMessageHandler (.NET 6) 
    dateFormat  X
    axisFormat %s
    todayMarker off
    section Baseline
    This PR (5292) - mean (261ms)  : 257, 264
     .   : milestone, 261,
    master - mean (258ms)  : 250, 265
     .   : milestone, 258,

    section CallTarget+Inlining+NGEN
    This PR (5292) - mean (857ms)  : 838, 877
     .   : milestone, 857,
    master - mean (855ms)  : 834, 876
     .   : milestone, 855,

```

File: tracer/src/Datadog.Trace/Iast/RangeList.cs [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+internal class RangeList
+{
+    private readonly int _maxRangeCount = Iast.Instance.Settings.MaxRangeCount;
+
+    private readonly Range[] _ranges;
+    private int _remaining;
+
+    internal RangeList(int count)
+    {
+        var finalCount = count > _maxRangeCount ? _maxRangeCount : count;
```
Comment (daniel-romano-DD) [inline tracer/src/Datadog.Trace/Iast/RangeList.cs:21]:
```suggestion
        var finalCount = Math.Min(count, _maxRangeCount);
```

Comment (daniel-romano-DD) [reply tracer/src/Datadog.Trace/Iast/RangeList.cs:21]: NIT: It's easier to understand IMO

**Review by daniel-romano-DD — APPROVED** [review]
Great, with a minor NIT. Thanks a lot

File: tracer/test/Datadog.Trace.Security.Unit.Tests/IAST/IastSettingsTests.cs [hunk 2]
```diff
@@ -70,6 +70,28 @@ public void GivenIastSettings_WhenSetMaxConcurrentRequestsToMinus1_MaxConcurrent
...
+        var iastSettings = new IastSettings(settings, NullConfigurationTelemetry.Instance);
+        Assert.Equal(5, iastSettings.MaxRangeCount);
+    }
+
+    [Fact]
+    public void GivenIastSettings_WhenSetMaxRangeCountToMinus1_MaxRangeCountIsDefaultValue()
+    {
+        var settings = new CustomSettingsForTests(new Dictionary<string, object>()
+        {
+            { ConfigurationKeys.Iast.MaxRangeCount, -1 }
```
Comment (lucaspimentel) [inline tracer/test/Datadog.Trace.Security.Unit.Tests/IAST/IastSettingsTests.cs:89]: I would also add a test for `0`, since the validation code doesn't allow zero.
Or maybe you could allow zero if it makes sense. Now sure what this is used for, so I can't say 😝

**Review by lucaspimentel — APPROVED** [review]

File: tracer/test/Datadog.Trace.Security.Unit.Tests/IAST/IastSettingsTests.cs (hunk 2 above)
Comment (e-n-0) [reply tracer/test/Datadog.Trace.Security.Unit.Tests/IAST/IastSettingsTests.cs:89]: I added a test for 0 then :)
